File: recipe-data/generate_recipes.py
import requests
import json
import time
from collections import OrderedDict

BASE_URL = "https://www.themealdb.com/api/json/v1/1"

OUTPUT_FILE = "recipes.json"

# We deliberately select from different categories.
# The script will collect recipes from these groups
# and stop when it reaches 50 unique recipes.

CATEGORY_LIMITS = {
    "Chicken": 6,
    "Beef": 4,
    "Vegetarian": 8,
    "Seafood": 5,
    "Pasta": 5,
    "Dessert": 5,
    "Breakfast": 5,
    "Lamb": 3,
    "Pork": 2,
    "Goat": 1,
    "Miscellaneous": 6
}

# ---------------------------------------------------------
# Helper functions
# ---------------------------------------------------------

def get_json(url, params=None):
    response = requests.get(url, params=params, timeout=15)
    response.raise_for_status()
    return response.json()


def get_category_meals(category):
    data = get_json(
        f"{BASE_URL}/filter.php",
        {"c": category}
    )

    return data.get("meals") or []


def get_recipe_details(meal_id):
    data = get_json(
        f"{BASE_URL}/lookup.php",
        {"i": meal_id}
    )

    meals = data.get("meals")

    if not meals:
        return None

    return meals[0]


def extract_ingredients(meal):
    ingredients = []

    for i in range(1, 21):
        ingredient = meal.get(f"strIngredient{i}")
        measure = meal.get(f"strMeasure{i}")

        if ingredient and ingredient.strip():
            ingredients.append({
                "name": ingredient.strip(),
                "quantity": measure.strip() if measure else ""
            })

    return ingredients


def clean_tags(tags):
    if not tags:
        return []

    return [
        tag.strip()
        for tag in tags.split(",")
        if tag.strip()
    ]


def derive_meal_type(category, title):
    category = category.lower()
    title = title.lower()

    if category == "dessert":
        return "Dessert"

    if category == "breakfast":
        return "Breakfast"

    if any(word in title for word in [
        "salad",
        "soup",
        "sandwich",
        "wrap"
    ]):
        return "Lunch"

    return "Dinner"


def derive_diet(category, title, ingredients):
    category = category.lower()
    title = title.lower()

    if category == "vegetarian":
        return "Vegetarian"

    meat_words = [
        "chicken",
        "beef",
        "pork",
        "lamb",
        "goat",
        "turkey",
        "duck",
        "fish",
        "salmon",
        "tuna",
        "prawn",
        "shrimp",
        "anchovy"
    ]

    ingredient_text = " ".join(
        item["name"].lower()
        for item in ingredients
    )

    combined = title + " " + ingredient_text

    if any(word in combined for word in meat_words):
        return "Non-Vegetarian"

    return "Vegetarian"


def derive_difficulty(ingredients, instructions):
    ingredient_count = len(ingredients)
    instruction_length = len(instructions or "")

    if ingredient_count <= 6 and instruction_length < 800:
        return "Easy"

    if ingredient_count <= 12 and instruction_length < 1600:
        return "Medium"

    return "Hard"


def derive_cook_time_minutes(title, difficulty):
    import re

    match = re.search(r"(\d+)\s*-?\s*(?:minute|min)\b", title, re.IGNORECASE)

    if match:
        return int(match.group(1))

    if difficulty == "Easy":
        return 25

    if difficulty == "Hard":
        return 60

    return 45


def derive_calories(meal_type, category):
    category_lower = category.lower()

    if meal_type == "Dessert" or category_lower == "dessert":
        return 350

    if meal_type == "Breakfast" or category_lower == "breakfast":
        return 400

    if category_lower in ("seafood", "vegetarian"):
        return 450

    return 550


def transform_recipe(meal):
    ingredients = extract_ingredients(meal)

    category = meal.get("strCategory") or "Miscellaneous"
    title = meal.get("strMeal") or "Unknown Recipe"
    instructions = meal.get("strInstructions") or ""
    meal_type = derive_meal_type(category, title)
    difficulty = derive_difficulty(ingredients, instructions)

    return {
        "externalId": meal.get("idMeal"),

        "title": title,

        "description": (
            f"{title} - a {category.lower()} recipe "
            f"from {meal.get('strArea') or 'international'} cuisine."
        ),

        "imageUrl": meal.get("strMealThumb"),

        "instructions": instructions.strip(),

        "cuisine": meal.get("strArea") or "International",

        "category": category,

        "mealType": meal_type,

        "diet": derive_diet(
            category,
            title,
            ingredients
        ),

        "difficulty": difficulty,

        "cookTimeMinutes": derive_cook_time_minutes(title, difficulty),

        "calories": derive_calories(meal_type, category),

        "ingredients": ingredients,

        "tags": clean_tags(
            meal.get("strTags")
        ),

        "youtubeUrl": meal.get("strYoutube"),

        "source": "TheMealDB",

        "isAiGenerated": False
    }


# ---------------------------------------------------------
# Main
# ---------------------------------------------------------

def main():

    recipes = OrderedDict()

    print("\n🍳 ChefMate AI Recipe Importer")
    print("-----------------------------------")

    for category, limit in CATEGORY_LIMITS.items():

        print(f"\nFetching category: {category}")

        try:
            meals = get_category_meals(category)

        except Exception as error:
            print(f"❌ Failed: {error}")
            continue

        count = 0

        for meal in meals:

            if len(recipes) >= 50:
                break

            meal_id = meal.get("idMeal")

            if not meal_id:
                continue

            if meal_id in recipes:
                continue

            try:
                print(
                    f"  → Fetching {meal.get('strMeal')}"
                )

                full_recipe = get_recipe_details(meal_id)

                if not full_recipe:
                    continue

                recipe = transform_recipe(full_recipe)

                recipes[meal_id] = recipe

                count += 1

                # Small delay so we're polite to the API.
                time.sleep(0.15)

                if count >= limit:
                    break

            except Exception as error:
                print(
                    f"  ⚠️ Skipping {meal_id}: {error}"
                )

        print(
            f"  ✓ Added {count} recipes"
        )

        if len(recipes) >= 50:
            break

    # -----------------------------------------------------
    # Save exactly 50 recipes
    # -----------------------------------------------------

    final_recipes = list(recipes.values())[:50]

    with open(
        OUTPUT_FILE,
        "w",
        encoding="utf-8"
    ) as file:

        json.dump(
            final_recipes,
            file,
            indent=2,
            ensure_ascii=False
        )

    print("\n-----------------------------------")
    print(
        f"✅ DONE! Created {len(final_recipes)} recipes."
    )
    print(
        f"📄 File: {OUTPUT_FILE}"
    )

    # -----------------------------------------------------
    # Print summary
    # -----------------------------------------------------

    print("\nRecipe distribution:")

    category_counts = {}

    for recipe in final_recipes:

        category = recipe["category"]

        category_counts[category] = (
            category_counts.get(category, 0) + 1
        )

    for category, count in category_counts.items():

        print(
            f"  {category}: {count}"
        )

    print("\n🎉 ChefMate recipe dataset ready!")


if __name__ == "__main__":
    main()

File: recipe-data/generate_recipes.mjs
import { writeFileSync } from "node:fs";

const BASE_URL = "https://www.themealdb.com/api/json/v1/1";
const OUTPUT_FILE = "recipes.json";

const CATEGORY_LIMITS = {
  Chicken: 6,
  Beef: 4,
  Vegetarian: 8,
  Seafood: 5,
  Pasta: 5,
  Dessert: 5,
  Breakfast: 5,
  Lamb: 3,
  Pork: 2,
  Goat: 1,
  Miscellaneous: 6,
};

async function getJson(url, params = {}) {
  const query = new URLSearchParams(params).toString();
  const fullUrl = query ? `${url}?${query}` : url;
  const response = await fetch(fullUrl, { signal: AbortSignal.timeout(15000) });

  if (!response.ok) {
    throw new Error(`HTTP ${response.status} for ${fullUrl}`);
  }

  return response.json();
}

async function getCategoryMeals(category) {
  const data = await getJson(`${BASE_URL}/filter.php`, { c: category });
  return data.meals ?? [];
}

async function getRecipeDetails(mealId) {
  const data = await getJson(`${BASE_URL}/lookup.php`, { i: mealId });
  return data.meals?.[0] ?? null;
}

function extractIngredients(meal) {
  const ingredients = [];

  for (let i = 1; i <= 20; i += 1) {
    const ingredient = meal[`strIngredient${i}`];
    const measure = meal[`strMeasure${i}`];

    if (ingredient?.trim()) {
      ingredients.push({
        name: ingredient.trim(),
        quantity: measure?.trim() ?? "",
      });
    }
  }

  return ingredients;
}

function cleanTags(tags) {
  if (!tags) {
    return [];
  }

  return tags
    .split(",")
    .map((tag) => tag.trim())
    .filter(Boolean);
}

function deriveMealType(category, title) {
  const categoryLower = category.toLowerCase();
  const titleLower = title.toLowerCase();

  if (categoryLower === "dessert") {
    return "Dessert";
  }

  if (categoryLower === "breakfast") {
    return "Breakfast";
  }

  if (["salad", "soup", "sandwich", "wrap"].some((word) => titleLower.includes(word))) {
    return "Lunch";
  }

  return "Dinner";
}

function deriveDiet(category, title, ingredients) {
  const categoryLower = category.toLowerCase();
  const titleLower = title.toLowerCase();

  if (categoryLower === "vegetarian") {
    return "Vegetarian";
  }

  const meatWords = [
    "chicken",
    "beef",
    "pork",
    "lamb",
    "goat",
    "turkey",
    "duck",
    "fish",
    "salmon",
    "tuna",
    "prawn",
    "shrimp",
    "anchovy",
  ];

  const ingredientText = ingredients.map((item) => item.name.toLowerCase()).join(" ");
  const combined = `${titleLower} ${ingredientText}`;

  if (meatWords.some((word) => combined.includes(word))) {
    return "Non-Vegetarian";
  }

  return "Vegetarian";
}

function deriveDifficulty(ingredients, instructions) {
  const ingredientCount = ingredients.length;
  const instructionLength = instructions?.length ?? 0;

  if (ingredientCount <= 6 && instructionLength < 800) {
    return "Easy";
  }

  if (ingredientCount <= 12 && instructionLength < 1600) {
    return "Medium";
  }

  return "Hard";
}

function deriveCookTimeMinutes(title, difficulty) {
  const match = title.match(/(\d+)\s*-?\s*(?:minute|min)\b/i);

  if (match) {
    return Number(match[1]);
  }

  if (difficulty === "Easy") {
    return 25;
  }

  if (difficulty === "Hard") {
    return 60;
  }

  return 45;
}

function deriveCalories(mealType, category) {
  const categoryLower = category.toLowerCase();

  if (mealType === "Dessert" || categoryLower === "dessert") {
    return 350;
  }

  if (mealType === "Breakfast" || categoryLower === "breakfast") {
    return 400;
  }

  if (categoryLower === "seafood" || categoryLower === "vegetarian") {
    return 450;
  }

  return 550;
}

function transformRecipe(meal) {
  const ingredients = extractIngredients(meal);
  const category = meal.strCategory ?? "Miscellaneous";
  const title = meal.strMeal ?? "Unknown Recipe";
  const instructions = meal.strInstructions ?? "";
  const mealType = deriveMealType(category, title);
  const difficulty = deriveDifficulty(ingredients, instructions);

  return {
    externalId: meal.idMeal,
    title,
    description: `${title} - a ${category.toLowerCase()} recipe from ${meal.strArea ?? "international"} cuisine.`,
    imageUrl: meal.strMealThumb,
    instructions: instructions.trim(),
    cuisine: meal.strArea ?? "International",
    category,
    mealType,
    diet: deriveDiet(category, title, ingredients),
    difficulty,
    cookTimeMinutes: deriveCookTimeMinutes(title, difficulty),
    calories: deriveCalories(mealType, category),
    ingredients,
    tags: cleanTags(meal.strTags),
    youtubeUrl: meal.strYoutube,
    source: "TheMealDB",
    isAiGenerated: false,
  };
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
  const recipes = new Map();

  console.log("\nChefMate AI Recipe Importer");
  console.log("-----------------------------------");

  for (const [category, limit] of Object.entries(CATEGORY_LIMITS)) {
    console.log(`\nFetching category: ${category}`);

    let meals;

    try {
      meals = await getCategoryMeals(category);
    } catch (error) {
      console.log(`Failed: ${error.message}`);
      continue;
    }

    let count = 0;

    for (const meal of meals) {
      if (recipes.size >= 50) {
        break;
      }

      const mealId = meal.idMeal;

      if (!mealId || recipes.has(mealId)) {
        continue;
      }

      try {
        console.log(`  -> Fetching ${meal.strMeal}`);

        const fullRecipe = await getRecipeDetails(mealId);

        if (!fullRecipe) {
          continue;
        }

        recipes.set(mealId, transformRecipe(fullRecipe));
        count += 1;

        await sleep(150);

        if (count >= limit) {
          break;
        }
      } catch (error) {
        console.log(`  Skipping ${mealId}: ${error.message}`);
      }
    }

    console.log(`  Added ${count} recipes`);

    if (recipes.size >= 50) {
      break;
    }
  }

  const finalRecipes = [...recipes.values()].slice(0, 50);

  writeFileSync(OUTPUT_FILE, JSON.stringify(finalRecipes, null, 2), "utf8");

  console.log("\n-----------------------------------");
  console.log(`DONE! Created ${finalRecipes.length} recipes.`);
  console.log(`File: ${OUTPUT_FILE}`);

  console.log("\nRecipe distribution:");

  const categoryCounts = finalRecipes.reduce((counts, recipe) => {
    counts[recipe.category] = (counts[recipe.category] ?? 0) + 1;
    return counts;
  }, {});

  for (const [category, count] of Object.entries(categoryCounts)) {
    console.log(`  ${category}: ${count}`);
  }

  console.log("\nChefMate recipe dataset ready!");
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
